internal/topology/variables/patch_validation.go

```diff
@@ -0,0 +1,222 @@
+
+		allErrs = append(allErrs,
+			validatePatchJSONPatches(definition.JSONPatches, variables, path)...)
+	}
+	return allErrs
+}
+
+func validatePatchJSONPatches(jsonPatches []clusterv1.JSONPatch, variables []clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	validOps := sets.NewString("add", "replace", "remove")
+	for i, jsonPatch := range jsonPatches {
+		if !validOps.Has(jsonPatch.Op) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Child("op").Index(i),
+					jsonPatch,
+					fmt.Sprintf("jsonPatch op must be one of \"add\" \"remove\" or \"replace\". %v is not valid.", jsonPatch.Op),
+				))
+		}
+
+		if !strings.HasPrefix(jsonPatch.Path, "/spec/") {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch path must start with \"/spec/\".",
+				))
+		}
+
+		if _, err := jsonptr.Parse(jsonPatch.Path); err != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Child("path").Index(i),
+					jsonPatch,
+					fmt.Sprintf("jsonPatch path must be a valid JSON pointer (RFC 6901). %s not valid.", jsonPatch.Path),
+				))
+		}
+
+		// return early if the jsonPatch does not have "replace" or "add" op. Additional validation is not needed.
+		if jsonPatch.Op != "add" && jsonPatch.Op != "replace" {
+			return allErrs
+		}
+
+		if jsonPatch.Value == nil && jsonPatch.ValueFrom == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch must define one of Value or ValueFrom.",
+				))
+		}
+		if jsonPatch.Value != nil && jsonPatch.ValueFrom != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch can not define both Value and ValueFrom.",
+				))
+		}
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template == nil && jsonPatch.ValueFrom.Variable == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"ValueFrom must set either \"/Template or Variable/\".",
+				))
+		}
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template != nil && jsonPatch.ValueFrom.Variable != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"ValueFrom can not set both \"/Template and Variable/\".",
+				))
+		}
+
+		//TODO: return early if the jsonPatch has the "replace" op. Additional validation is not needed.
+		// if jsonPatch.Op == "replace" {
+		//	return allErrs
+		//}
+
+		// TODO: only the indexes `0` (prepend) and `-` (append) in combination with the `add` operation are allowed (append and prepend are the only allowed array modifications).
+		if strings.Contains(jsonPatch.Path, "ARRAY INDEXING") {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					fmt.Sprintf("Arrays can only be accessed using [0] (prepend) or [-] (append). %s not valid", "WHAT IS THIS"),
+				))
+		}
```

Comment:
Is this completed? the TODO, ARRAY INDEXING, WHAT IS THIS makes me think this bit is still WIP

Reply:
Yeah there's a few outstanding TODOs in the code. I'm working on getting them done now but there's outstanding issues on:
1) where this code should go (I'm happy with the webhook folder which is a good suggestion)
2) How to get a canonical list of defined builtin variables.

The other TODOs and comments I should have addressed in a commit pretty soon.

Reply:
Will be cleaned up :)

Reply:
> where this code should go (I'm happy with the webhook folder which is a good suggestion)

Fine for me too, if there is no reason why we need it accessible from somewhere else